bee-network/bee-autopeering/src/peer/peer_id.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Peer identities.
+
+use crate::hash;
+
+use crypto::signatures::ed25519::{PublicKey, SecretKey as PrivateKey, PUBLIC_KEY_LENGTH};
+use serde::{
+    de::{SeqAccess, Visitor},
+    ser::SerializeStruct,
+    Deserialize, Serialize,
+};
+
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+};
+
+const DISPLAY_LENGTH: usize = 16;
+
+/// Represents the unique identity of a peer in the network.
+#[derive(Copy, Clone)]
+pub struct PeerId {
+    // An ED25519 public key.
+    public_key: PublicKey,
+    // The `SHA256` hash of the public key.
+    id_bytes: [u8; hash::SHA256_LEN],
+}
+
+impl PeerId {
+    /// Creates a new peer identity.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Creates a peer identity from an ED25519 public key.
+    pub fn from_public_key(public_key: PublicKey) -> Self {
+        let id_bytes = hash::sha256(public_key.as_ref());
+
+        Self { id_bytes, public_key }
+    }
+
+    /// Returns the public key associated with this identity.
+    pub fn public_key(&self) -> &PublicKey {
+        &self.public_key
+    }
+
+    /// Creates the corresponding `libp2p_core::PublicKey` of this identity.
+    pub fn to_libp2p_public_key(&self) -> libp2p_core::PublicKey {
+        libp2p_core::PublicKey::Ed25519(
+            libp2p_core::identity::ed25519::PublicKey::decode(self.public_key.as_ref())
+                .expect("error decoding ed25519 public key from bytes"),
+        )
+    }
+
+    /// Creates the corresponding `libp2p_core::PeerId` of this identity.
+    pub fn to_libp2p_peer_id(&self) -> libp2p_core::PeerId {
+        libp2p_core::PeerId::from_public_key(self.to_libp2p_public_key())
+    }
+
+    /// Returns the actual bytes representing this id.
+    pub fn id_bytes(&self) -> &[u8; hash::SHA256_LEN] {
+        &self.id_bytes
+    }
+}
+
+impl Default for PeerId {
```

Comment:
I feel like Default should be deterministic, I wouldn't use Default to generate random stuff, I'd have an explicit API for that.

Reply:
I agree. I removed the Default impl. The API on PeerId now has a `generate` function (consistent with the underlying PrivateKey impl). I had to remove `new` also, so that clippy doesn't complain about "new_without_default".